File: refactoring-practice-1/src/main/java/dojo/introduce_null_object/Site.java
package dojo.introduce_null_object;

/*
 * A null vizsgálatok helyett hozzunk létre egy NullCustomer osztályt, és az adjon vissza olyan értékeket, amiket
 * a null vizsgálatok eredményei adtak vissza.
 * Lépések:
 * - NullCustomer osztály és Customer.isNull metódus létrehozása
 * - NullCustomer példány és Customer.isNull metódusok behelyettesítése
 * - Feltételes utasítások megszüntetése, NullCustomer metódusainak implementálásával
 */
public class Site {

	private Customer customer;
	
	public void setCustomer(Customer customer) {
		this.customer = customer;
	}
	
	public String getCustomerName() {
		if (customer == null) {
			return "occupant";
		} else {
			return customer.getName();
		}
	}
	
	public BillingPlan getBillingPlan() {
		if (customer == null) {
			return BillingPlan.BASIC;
		} else {
			return customer.getBillingPlan();
		}
	}
	
	public int getWeeksDelinquent() {
		if (customer == null) {
			return 0;
		} else {
			return customer.getBillingHistory().getWeeksDelinquentInLastYear();
		}
	}
}

File: refactoring-practice-1/src/main/typescript/introduce_null_object/site.ts
/*
 * A null vizsgálatok helyett hozzunk létre egy NullCustomer osztályt, és az adjon vissza olyan értékeket, amiket
 * a null vizsgálatok eredményei adtak vissza.
 */
import {Customer} from "./customer.js";
import {BillingPlan} from "./billing-plan.js";

export class Site {

    private customer: Customer;

    public setCustomer(customer: Customer) {
        this.customer = customer;
    }

    public getCustomerName(): string {
        if (this.customer == null) {
            return "occupant";
        } else {
            return this.customer.getName();
        }
    }

    public getBillingPlan(): BillingPlan {
        if (this.customer == null) {
            return BillingPlan.BASIC;
        } else {
            return this.customer.getBillingPlan();
        }
    }

    public getWeeksDelinquent(): number {
        if (this.customer == null) {
            return 0;
        } else {
            return this.customer.getBillingHistory().getWeeksDelinquentInLastYear();
        }
    }
}
